src/compiler/crystal/config.cr

```diff
@@ -34,8 +34,40 @@ module Crystal
       Time.unix(time).to_s("%Y-%m-%d")
     end
 
-    def self.default_target_triple
-      {{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple
+    @@default_target : Crystal::Codegen::Target?
+
+    def self.default_target : Crystal::Codegen::Target
+      @@default_target ||= begin
+        target = Crystal::Codegen::Target.new({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple)
+
+        if target.linux?
+          # The linux binary runs as well on linux-gnu as linux-musl, but the
+          # default target needs to match the C library used on the current
+          # system.
+          # This can be automatically detected from the output of `ldd --version`
+          # in order to use the appropriate environment target.
+          default_libc = target.gnu? ? "-gnu" : "-musl"
+
+          target = Crystal::Codegen::Target.new(target.to_s.sub(default_libc, "-#{runtime_libc}"))
+        end
+
+        target
+      end
+    end
+
+    private def self.runtime_libc
+      ldd_version = String.build do |io|
+        Process.new("ldd", {"--version"}, output: io, error: io).wait
```

Comment:
Why not `Process.run` directly?

Reply:
Probably just overlooked that. Maybe I did something else in the beginning...